pkg/middleware/auth/auth.go

```diff
@@ -341,11 +362,156 @@ func getSubjectName(user *ent.User) string {
 	return subjectName
 }
 
-// unauthorized returns a 401 Unauthorized response with the error message.
-func unauthorized(c echo.Context, err error) error {
-	if err := c.JSON(http.StatusUnauthorized, rout.ErrorResponse(err)); err != nil {
-		return err
+// unauthorized returns a 401 Unauthorized response with the error message
+func unauthorized(c echo.Context, err error, conf *Options, v tokens.Validator) error {
+	reqCtx := c.Request().Context()
+
+	if v != nil && conf != nil && conf.DBClient != nil {
+		orgID := orgIDFromToken(c, v)
+		if orgID != "" {
+			userID := userIDFromToken(c, v)
+			if userID != "" {
+				enforced, dbErr := isSSOEnforcedFunc(reqCtx, conf.DBClient, orgID)
+
+				if dbErr == nil && enforced {
+					role, mErr := orgRoleFunc(reqCtx, conf.DBClient, userID, orgID)
+
+					if mErr == nil && role != enums.RoleOwner {
+						if redirErr := c.Redirect(http.StatusFound, sso.SSOLogin(c.Echo(), orgID)); redirErr != nil {
+							return redirErr
+						}
+
+						return nil
+					}
+				}
+			}
+		}
+	}
+
+	if jsonErr := c.JSON(http.StatusUnauthorized, rout.ErrorResponse(err)); jsonErr != nil {
+		log.Error().Err(jsonErr).Msg("failed to write unauthorized JSON response")
+		return jsonErr
+	}
+
+	return nil
+}
+
+// orgIDFromToken extracts the organization ID from the token in the request context
+func orgIDFromToken(c echo.Context, v tokens.Validator) string {
+	authHeader := c.Request().Header.Get("Authorization")
+
+	if authHeader != "" {
+		token, err := auth.GetBearerToken(c)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to get bearer token from context")
+			return ""
+		}
+
+		if token != "" {
+			if claims, parseErr := v.Parse(token); parseErr == nil {
+				return claims.OrgID
+			}
+		}
+	}
+
+	cookie, err := c.Cookie("refresh_token")
+	if err == nil && cookie != nil && cookie.Value != "" {
+		refresh := cookie.Value
+
+		if claims, parseErr := v.Parse(refresh); parseErr == nil {
+			return claims.OrgID
+		}
+	}
+
+	return ""
+}
+
+// isSSOEnforced checks if SSO is enforced for the given organization ID
+func isSSOEnforced(ctx context.Context, db *ent.Client, orgID string) (bool, error) {
+	setting, err := db.OrganizationSetting.Query().Where(organizationsetting.OrganizationID(orgID)).
+		Only(privacy.DecisionContext(ctx, privacy.Allow))
+	if err != nil {
```

Comment:
You should just pull the field you need
```suggestion
	setting, err := db.OrganizationSetting.Query().Where(organizationsetting.OrganizationID(orgID)).
	        Select(organizationsetting.IdentityProviderLoginEnforced).
		Only(privacy.DecisionContext(ctx, privacy.Allow))
	if err != nil {
```